kernel/src/fs/cgroupfs/controller/mod.rs

```diff
@@ -0,0 +1,393 @@
+        cgroup_node: &dyn CgroupSysNode,
+    ) -> Result<usize> {
+        self.as_subcontrol()
+            .read_attr_at(name, offset, writer, cgroup_node)
+    }
+
+    fn write_attr(
+        &self,
+        name: &str,
+        reader: &mut VmReader,
+        cgroup_node: &dyn CgroupSysNode,
+    ) -> Result<usize> {
+        self.as_subcontrol().write_attr(name, reader, cgroup_node)
+    }
+}
+
+bitflags! {
+    /// Bitflags representing active/deactive sub-control state.
+    pub(super) struct SubCtrlState: u8 {
+        const MEMORY_CTRLS = 1 << 0;
+        const CPUSET_CTRLS = 1 << 1;
+        const PIDS_CTRLS = 1 << 2;
+    }
+}
+
+impl SubCtrlState {
+    pub(super) fn control_bit(name: &str) -> Option<Self> {
+        match name {
+            "memory" => Some(Self::MEMORY_CTRLS),
+            "cpuset" => Some(Self::CPUSET_CTRLS),
+            "pids" => Some(Self::PIDS_CTRLS),
+            _ => None,
+        }
+    }
+
+    /// Checks if a sub-control is active in the current state.
+    ///
+    /// If the given name does not represent a supported controller,
+    /// returns `None`.
+    pub(super) fn is_active(&self, name: &str) -> Option<bool> {
+        Self::control_bit(name).map(|bit| self.contains(bit))
+    }
+
+    fn activate(&mut self, name: &str) {
+        if let Some(bit) = Self::control_bit(name) {
+            *self |= bit;
+        }
+    }
+
+    fn deactivate(&mut self, name: &str) {
+        if let Some(bit) = Self::control_bit(name) {
+            *self -= bit;
+        }
+    }
+
+    pub(super) fn show(&self) -> String {
+        let mut controllers = Vec::new();
+
+        if self.contains(Self::MEMORY_CTRLS) {
+            controllers.push("memory");
+        }
+        if self.contains(Self::CPUSET_CTRLS) {
+            controllers.push("cpuset");
+        }
+        if self.contains(Self::PIDS_CTRLS) {
+            controllers.push("pids");
+        }
+
+        controllers.join(" ")
+    }
+}
+
+/// The controller for a single cgroup.
+///
+/// This struct can manage the activation state of each sub-control, and dispatches read/write
+/// operations to the appropriate sub-controllers.
+///
+/// The following is an explanation of the activation for sub-controls and controllers.
+/// When a cgroup activates a specific sub-control (e.g., memory, io), it means this control
+/// capability is being delegated to its children. Consequently, the corresponding controller
+/// within the child nodes will be activated.
+///
+/// The root node serves as the origin for all these control capabilities, so the controllers
+/// it possesses are always active. For any other node, only if its parent node first enables
+/// a sub-control, its corresponding controller will be activated.
+///
+/// Among all nodes, the fundamental cgroup controller is always active.
+pub(super) struct Controller {
+    sub_ctrl_state: Mutex<SubCtrlState>,
+    controllers: BTreeMap<SysStr, RcuOption<Arc<SubController>>>,
```

Comment:
I will prefer:
```rust
pub(super) struct Controller {
    sub_ctrl_state: Mutex<SubCtrlState>,

    memory: RcuOption<MemoryController>,
    cpuset: RcuOption<CpuSetController>,
    pids: RcuOption<PidsController>,
}
```

Also, we should have an enum to replace some `name: &str` argument.
```rust
enum ControllerType {
    Memory,
    CpuSet,
    Pids,
}
```